src/experimental-layers/src/text-layer/text-layer.js

```diff
@@ -0,0 +1,155 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import {CompositeLayer} from 'deck.gl';
+import MultiIconLayer from './multi-icon-layer/multi-icon-layer';
+import {makeFontAtlas} from './font-atlas';
+
+const DEFAULT_COLOR = [0, 0, 0, 255];
+const TEXT_ANCHOR = {
+  start: 1,
+  middle: 0,
+  end: -1
+};
+const ALIGNMENT_BASELINE = {
+  top: 1,
+  center: 0,
+  bottom: -1
+};
+// currently the font family is invisible to the user
+const FONT_FAMILY = '"Lucida Console", Monaco, monospace';
+
+const defaultProps = {
+  getText: x => x.text,
+  getPosition: x => x.coordinates,
+  getColor: x => x.color || DEFAULT_COLOR,
+  getSize: x => x.size || 32,
+  getAngle: x => x.angle || 0,
+  getTextAnchor: x => x.textAnchor || 'middle',
+  getAlignmentBaseline: x => x.alignmentBaseline || 'center',
+  getPixelOffset: x => x.pixelOffset || [0, 0],
+  fp64: false
+};
+
+export default class TextLayer extends CompositeLayer {
+  initializeState() {
+    const {gl} = this.context;
+    const {mapping, texture} = makeFontAtlas(gl, FONT_FAMILY);
+    this.state = {
+      iconAtlas: texture,
+      iconMapping: mapping
+    };
+  }
+
+  shouldUpdateState({changeFlags}) {
+    return changeFlags.somethingChanged;
+  }
+
+  updateState({props, oldProps, changeFlags}) {
+    if (changeFlags.dataChanged) {
+      this.transformStringToLetters();
+    }
+  }
+
+  transformStringToLetters() {
+    const {data, getText, getPosition} = this.props;
+    if (!data || data.length === 0) {
+      return;
+    }
+
+    const transformedData = data
+      .map(val => {
+        const text = getText(val);
+        const letters = Array.from(text);
+        const position = getPosition(val);
+        if (!text) {
+          return [];
+        }
+        return letters.map((letter, i) =>
+          Object.assign({}, val, {text: letter, position, index: i, len: text.length})
+        );
+      })
+      .reduce((prev, curr) => [...prev, ...curr]);
+
+    this.setState({data: transformedData});
+  }
+
+  getAnchorXFromTextAnchor(textAnchor) {
+    if (!TEXT_ANCHOR.hasOwnProperty(textAnchor)) {
+      throw new Error(`Invalid text anchor parameter: ${textAnchor}`);
+    }
+    return TEXT_ANCHOR[textAnchor];
+  }
+
+  getAnchorYFromAlignmentBaseline(alignmentBaseline) {
+    if (!ALIGNMENT_BASELINE.hasOwnProperty(alignmentBaseline)) {
+      throw new Error(`Invalid alignment baseline parameter: ${alignmentBaseline}`);
+    }
+    return ALIGNMENT_BASELINE[alignmentBaseline];
+  }
+
+  renderLayers() {
+    const {data, iconAtlas, iconMapping} = this.state;
+
+    if (!iconMapping || !iconAtlas || !data) {
+      return null;
+    }
+
+    const {
+      getColor,
+      getSize,
+      getAngle,
+      getTextAnchor,
+      getAlignmentBaseline,
+      getPixelOffset,
+      fp64
+    } = this.props;
+
+    return [
+      new MultiIconLayer(
```

Comment:
is it the plan that we remove `MultiIconLayer` altogether and merge its content to the text layer? since its usage is pretty much limited to render texts?

Reply:
Discussed this with @ibgreen. We think to keep it separate for now in case there are other use cases for multiple icons. Or we can also move it to somewhere that contains a collection of "helper" classes/functions for deck.gl (but not in the 4.2 release).

Reply:
okay

Reply:
We have a growing number of these internal layers in deck.gl (MultiIconLayer, SolidPolygonLayer, HexagonCellLayer etc). Not sure how to handle them, what directory to put them in etc. Would be a good API Audit topic.